backend/qemu.pm

```diff
@@ -355,68 +404,138 @@ sub save_storage_drives {
 sub save_snapshot {
     my ($self, $args) = @_;
     my $vmname = $args->{name};
-    my $rsp    = $self->handle_qmp_command(_wrap_hmc("savevm $vmname"));
-    if (!$rsp || $rsp->{return} ne '') {
-        die "Could not save snapshot \'$vmname\' " . Dumper($rsp);
-    }
+    my $bdc    = $self->{proc}->blockdev_conf;
+
+    my $rsp = $self->handle_qmp_command({execute => 'query-status'}, fatal => 1);
+    bmwqemu::diag("Saving snapshot (Current VM state is $rsp->{return}->{status}).");
+    my $was_running = $rsp->{return}->{status} eq 'running';
+    $self->freeze_vm() if $was_running;
+
+    $self->save_console_snapshots($vmname);
+
+    my $snapshot = $self->{proc}->snapshot_conf->add_snapshot($vmname);
+    $bdc->for_each_drive(sub {
+            local $Data::Dumper::Indent   = 0;
+            local $Data::Dumper::Terse    = 1;
+            local $Data::Dumper::Sortkeys = 1;
+            my $drive = shift;
+
+            my $overlay = $bdc->add_snapshot_to_drive($drive, $snapshot);
+            my $req = {execute => 'blockdev-snapshot-sync',
+                arguments => {'node-name' => $overlay->backing_file->node_name,
+                    'snapshot-node-name' => $overlay->node_name,
+                    'snapshot-file'      => $overlay->file,
+                    format               => $overlay->driver}};
+            $rsp = $self->handle_qmp_command($req);
+
+            # Assumes errors are caused by pflash drives using an autogenerated
+            # blockdev node-name. Try again using the device id instead.
+            if ($rsp->{error}) {
+                diag('blockdev-snapshot-sync(' . Dumper($req) . ') -> ' . Dumper($rsp));
+                delete($req->{arguments}->{'node-name'});
+                $req->{arguments}->{device} = $overlay->backing_file->node_name;
+                $rsp = $self->handle_qmp_command($req);
+            }
+
+            diag('blockdev-snapshot-sync(' . Dumper($req) . ') -> ' . Dumper($rsp));
+    });
+
+    mkpath(VM_SNAPSHOTS_DIR);
+    $self->_migrate_to_file(filename => VM_SNAPSHOTS_DIR . '/' . $snapshot->name,
+        compress_level => 9);
+    diag('Snapshot complete');
+
+    $self->cont_vm() if $was_running;
     return;
 }
 
 sub load_snapshot {
     my ($self, $args) = @_;
     my $vmname = $args->{name};
-    my $rsp    = $self->handle_qmp_command(_wrap_hmc("loadvm $vmname"));
-    if (!$rsp || $rsp->{return} ne '') {
-        die "Could not load snapshot \'$vmname\': " . Dumper($rsp);
-    }
-    return $rsp;
+
+    my $rsp = $self->handle_qmp_command({execute => 'query-status'}, fatal => 1);
+    bmwqemu::diag("Loading snapshot (Current VM state is $rsp->{return}->{status}).");
+    my $was_running = $rsp->{return}->{status} eq 'running';
```

Comment:
Just a bit of reminder (for myself too) - this could be problematic with MM jobs if they are loading snapshots and using openvswitch. But i'm not aware we are using it in such way nor how does qemu handles networking devices attached to it in this case (state after restart, etc.)

Reply:
Yeah, I think maybe snapshots should be disabled for VM's which are doing anything with external tools like Openvswitch. The networking code needs to support saving and loading state as well. So we can roll back when loading a snapshot. It might be a good idea to move the networking configuration into the object model as well, especially as a number of people doing networking tests want way more advanced network configurations.